server/src/main/java/io/unitycatalog/server/persist/ModelRepository.java

```diff
@@ -345,17 +358,426 @@ public void deleteRegisteredModel(String fullName, boolean force) {
+  }
+
+  public List<ModelVersionInfoDAO> getModelVersionsDao(
+      Session session, UUID registeredModelId, String token, int maxResults) {
+    String hql =
+        "FROM ModelVersionInfoDAO t WHERE t.registeredModelId = :registeredModelId AND t.version > :token ORDER BY t.version ASC";
+    Query<ModelVersionInfoDAO> query = session.createQuery(hql, ModelVersionInfoDAO.class);
+    query.setParameter("registeredModelId", registeredModelId);
+    query.setParameter("token", Long.parseLong(token));
+    query.setMaxResults(maxResults);
+    LOGGER.info("Finding model versions by registeredModelId: " + registeredModelId);
+    return query.getResultList(); // Returns null if no result is found
+  }
+
+  public ModelVersionInfo createModelVersion(CreateModelVersion createModelVersion) {
+    long createTime = System.currentTimeMillis();
+    String modelVersionId = UUID.randomUUID().toString();
+    String catalogName = createModelVersion.getCatalogName();
+    String schemaName = createModelVersion.getSchemaName();
+    String modelName = createModelVersion.getModelName();
+    ModelVersionInfo modelVersionInfo =
+        new ModelVersionInfo()
+            .modelVersionId(modelVersionId)
+            .modelName(createModelVersion.getModelName())
+            .catalogName(createModelVersion.getCatalogName())
+            .schemaName(createModelVersion.getSchemaName())
+            .runId(createModelVersion.getRunId())
+            .source(createModelVersion.getSource())
+            .status(ModelVersionStatus.PENDING_REGISTRATION)
+            .comment(createModelVersion.getComment())
+            .createdAt(createTime)
+            .updatedAt(createTime);
+    String registeredModelFullName = getRegisteredModelFullName(catalogName, schemaName, modelName);
+    LOGGER.info("Creating Registered Model: " + registeredModelFullName);
+
+    Transaction tx;
+    try (Session session = SESSION_FACTORY.openSession()) {
+      UUID catalogId = getCatalogId(session, catalogName);
+      UUID schemaId = getSchemaId(session, catalogName, schemaName);
+      tx = session.beginTransaction();
+      try {
+        // Check if registered model already exists
+        RegisteredModelInfoDAO existingRegisteredModel =
+            getRegisteredModelDao(session, schemaId, modelName);
+        if (existingRegisteredModel == null) {
+          throw new BaseException(ErrorCode.NOT_FOUND, "Registered model not found: " + modelName);
+        }
+        if (existingRegisteredModel.getMaxVersionNumber() == null
+            || existingRegisteredModel.getMaxVersionNumber() < 0) {
+          throw new BaseException(
+              ErrorCode.OUT_OF_RANGE,
+              "Registered model has invalid max model version: "
+                  + existingRegisteredModel.getMaxVersionNumber());
+        }
+        UUID modelId = existingRegisteredModel.getId();
+        Long version = existingRegisteredModel.getMaxVersionNumber() + 1;
+        String storageLocation =
```

Comment:
Will add a follow up commit that deals with storage location creation here and in registered models, as well as rollback of the creation on failure.